metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -45,71 +49,183 @@
+        }
+
+        if (newRange.max() < newRange.min()) {
             return new ApiError(Errors.INVALID_UPDATE_VERSION,
-                "The lower value for the new range cannot be less than 1.");
+                "The upper value for the new range cannot be less than the lower value.");
         }
+
         if (newRange.max() <= 0) {
             return new ApiError(Errors.INVALID_UPDATE_VERSION,
                 "The upper value for the new range cannot be less than 1.");
         }
-        VersionRange localRange = supportedFeatures.get(featureName);
-        if (localRange == null || !localRange.contains(newRange)) {
+
+        if (!canSupportVersion(featureName, newRange)) {
             return new ApiError(Errors.INVALID_UPDATE_VERSION,
                 "The controller does not support the given feature range.");
         }
-        for (Entry<Integer, Map<String, VersionRange>> brokerEntry :
-            brokerFeatures.entrySet()) {
+
+        for (Entry<Integer, Map<String, VersionRange>> brokerEntry : brokersAndFeatures.entrySet()) {
             VersionRange brokerRange = brokerEntry.getValue().get(featureName);
             if (brokerRange == null || !brokerRange.contains(newRange)) {
                 return new ApiError(Errors.INVALID_UPDATE_VERSION,
                     "Broker " + brokerEntry.getKey() + " does not support the given " +
                         "feature range.");
             }
         }
-        VersionRange currentRange = finalizedVersions.get(featureName);
+
         if (currentRange != null && currentRange.max() > newRange.max()) {
             if (!downgradeable) {
                 return new ApiError(Errors.INVALID_UPDATE_VERSION,
                     "Can't downgrade the maximum version of this feature without " +
                     "setting downgradable to true.");
             }
         }
-        records.add(new ApiMessageAndVersion(
-            new FeatureLevelRecord().setName(featureName).
+
+        if (featureName.equals(MetadataVersions.FEATURE_NAME)) {
+            // Perform additional checks if we're updating metadata.version
+            return updateMetadataVersion(newRange.max(), metadataVersionProvider.activeVersion().version(), records::add);
+        } else {
+            records.add(new ApiMessageAndVersion(
+                new FeatureLevelRecord().setName(featureName).
+                    setMinFeatureLevel(newRange.min()).setMaxFeatureLevel(newRange.max()),
+                metadataVersionProvider.activeVersion().recordVersion(FEATURE_LEVEL_RECORD)));
+            return ApiError.NONE;
+        }
+    }
+
+    /**
+     * Perform some additional validation for metadata.version updates.
+     */
+    private ApiError updateMetadataVersion(short newVersion,
+                                           short recordVersion,
+                                           Consumer<ApiMessageAndVersion> recordConsumer) {
```

Comment:
Notice here we need the desired metadata.version as well as the FeatureLevelRecord version to write. This is to handle the initialization case.